File: morkato-api/database/src/test/java/org/morkato/api/repository/DatabaseGuildRepositoryTest.java
package org.morkato.api.repository;

import jakarta.validation.Validation;
import jakarta.validation.Validator;
import org.flywaydb.core.Flyway;
import org.jetbrains.exposed.sql.Database;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.morkato.api.repository.guild.GuildRepository;
import org.morkato.database.DatabaseProvider;
import org.morkato.database.repository.DatabaseGuildRepository;

import javax.sql.DataSource;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class DatabaseGuildRepositoryTest extends GuildRepositoryTest {
  private DataSource datasource;
  private Flyway flyway;
  private Database database;
  private Validator validator;
  @Override
  protected GuildRepository getRepository() {
    datasource = DatabaseProvider.Companion.getDatasource(
      "morkato",
      "morkato",
      "jdbc:postgresql://localhost:8080/morkato-dev",
      "org.postgresql.Driver"
    );
    flyway = DatabaseProvider.Companion.getFlyway(datasource)
      .cleanDisabled(false)
      .load();
    flyway.clean();
    flyway.migrate();
    Database database = DatabaseProvider.Companion.getDatabase(datasource);
    Validator validator = Validation.buildDefaultValidatorFactory().getValidator();
    return new DatabaseGuildRepository(database, validator);
  }

  @Override
  protected void clean() {
    flyway.clean();
    flyway.migrate();
  }
}

File: morkato-api-database/src/test/kotlin/org/morkato/database/repository/DatabaseGuildRepositoryTest.kt
package org.morkato.database.repository

import jakarta.validation.Validation
import jakarta.validation.Validator
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.TestInstance
import org.morkato.api.repository.GuildRepository
import org.morkato.api.repository.GuildRepositoryTest
import org.morkato.database.DatabaseProvider
import org.morkato.utility.MorkatoConfigLoader
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class DatabaseGuildRepositoryTest : GuildRepositoryTest() {
  companion object {
    const val THREAD_POOL_SIZE: Int = 5
  }
  private val executor: ExecutorService
  private val settings: Properties
  private val provider: DatabaseProvider
  private val validator: Validator
  init {
    settings = MorkatoConfigLoader.loadDefault()
    executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE)
    provider = DatabaseProvider.get(settings)
    validator = Validation.buildDefaultValidatorFactory().getValidator()
  }

  override fun cleanRepository() {
    provider.clean()
    provider.migrate()
  }

  override fun snapshotRepository() {
    
  }

  override fun createRepository() : GuildRepository {
    return DatabaseGuildRepository(executor, provider.getDatabase(), validator)
  }

  @AfterAll
  fun closeThreadPool() {
    println("Fechando o ExecutorService")
    this.executor.shutdown()
  }

//  @Test
//  @DisplayName("Testando erro de validação de ID da guilda.")
//  fun guildIdIsInvalid() {
//    this.testGuildIdIsInvalid()
//  }
//
//  @Test
//  @DisplayName("Testando erro de ID não existente")
//  fun guildIsNotFound() {
//    this.testGuildNotFound()
//  }
}